dali/pipeline/operator/direct_operator.h

```diff
@@ -0,0 +1,241 @@
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+    cudaStream_t cuda_stream) {
+  ws.Clear();
+  ws.set_stream(cuda_stream);
+  CUDA_CALL(cudaStreamSynchronize(cuda_stream));
+  auto output = RunImpl<CPUBackend, GPUBackend, TensorVector<CPUBackend>, TensorList<GPUBackend>>(
+      inputs, kwargs);
+  CUDA_CALL(cudaStreamSynchronize(cuda_stream));
+  return output;
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<CPUBackend>>> DirectOperator<CPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<CPUBackend, CPUBackend>(inputs, kwargs, shared_thread_pool.get());
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> DirectOperator<GPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<GPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<GPUBackend, GPUBackend>(inputs, kwargs, shared_cuda_stream);
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> DirectOperator<MixedBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<CPUBackend, GPUBackend>(inputs, kwargs, shared_cuda_stream);
+}
+
+template <typename Backend>
+template <typename InBackend, typename OutBackend, typename WSInputType, typename WSOutputType>
+std::vector<std::shared_ptr<TensorList<OutBackend>>> DirectOperator<Backend>::RunImpl(
+    const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  // Convert and add inputs to the workspace.
+  for (size_t in_idx = 0; in_idx < inputs.size(); ++in_idx) {
+    auto tensor_in = std::make_shared<WSInputType>();
+    tensor_in->ShareData(*inputs[in_idx]);
+
+    // Set default layout for input if not specified.
+    if (tensor_in->GetLayout().empty()) {
+      auto default_layout = op_spec.GetSchema().GetInputLayout(
+          in_idx, tensor_in->shape().sample_dim(), tensor_in->GetLayout());
+      if (!default_layout.empty()) {
+        tensor_in->SetLayout(default_layout);
+      }
+    }
+
+    ws.AddInput(tensor_in);
+  }
+
+  for (auto &arg : kwargs) {
+    ws.AddArgumentInput(arg.first, arg.second);
+  }
+
+  std::vector<OutputDesc> output_desc{};
+  std::vector<std::shared_ptr<TensorList<OutBackend>>> outputs{};
+
+  outputs.reserve(num_outputs);
+
+  for (size_t i = 0; i < num_outputs; ++i) {
+    ws.AddOutput(std::make_shared<WSOutputType>(batch_size));
+  }
+
+  ws.SetBatchSizes(batch_size);
+
+  // Setup outputs.
+  if (op->Setup(output_desc, ws) && op->CanInferOutputs()) {
```

Comment:
```suggestion
  if (op->CanInferOutputs() && op->Setup(output_desc, ws)) {
```
If `op->CanInferOutputs()` return false we don't need to call `Setup`, I guess.

Reply:
done

Reply:
We do have to call `Setup`. Not calling it resulted in segfault (for `readers.File` but most likely same thing would happen for other operators).

Reply:
Yes. I'm talking rubbish. Sorry.